File: nodejs/node.js
var sys = require('sys');
var extend = require('./util');
var Thistle = require('./thistle');

function mark_safe(s) {
    s.is_safe = true;
    return s;
}

//
//  
//
var Node = function() {
    this.must_be_first = false;
    this.child_nodelists = ['nodelist'];
}

extend(Node, Object, {
    render : function (context) {
        // Return the node rendered as a string
    },
    get_nodes_by_type : function (nodetype) {
        var nodes = [];
        if (this instanceof nodetype)
            nodes.push(this); 
        for (var nidx = 0; nidx < this.child_nodelists.length; nidx++) {
            var nlist = this[this.child_nodelists[nidx]] || null;

            if (nlist != null) {
                (nlist.get_nodes_by_type(nodetype) || []).map(function (n) { nodes.push(n); });
            }
        }

        return nodes;
    }
});

//
//  
//
var NodeList = function() {
    this.contains_nontext = false;
    Node.apply(this, []);
}

extend(NodeList, Array, {
    render_node : function(node, context) {
        return node.render(context);
    },
        
    render : function(context) {
        var t = this;
        var bits = this.map(function(node) {
            if (node instanceof Node) {
                return t.render_node(node, context);
            } else {
                return node;
            }
        });

        return mark_safe(bits.join(''));
    },

    get_nodes_by_type : function(nodetype) {
        var nodes = []
        for (var idx = 0; idx < this.length; idx++) {
            (this[idx].get_nodes_by_type(nodetype) || []).map(function (n) { nodes.push(n); });
        }
        return nodes;
    },

    toString : function() {
        str = "<NodeList contains_nontext="+this.contains_nontext+" len="+this.length+" [";
        for (var i = 0; i < this.length; i++) {
            str += (i != 0 ? ", " : "");
            str += this[i];
        }
        return str + "]>";
    },
});

//
//
//
function TextNode(s) {
    Node.apply(this, []);
    this.s = s;
    this.must_be_first = false;
}
extend(TextNode, Node, {
    toString : function() {
        return "<Text Node: '" + this.s.substr(0,25) + "'>";
    },
    render : function(context) {
        return this.s;
    }
});

function VariableNode(filter_expression) {
    Node.apply(this, []);
    this.filter_expression = filter_expression;
    this.must_be_first = false;
}
extend(VariableNode, Node, {
    toString : function() {
        return "<Variable Node: " + this.filter_expression + ">";
    },
    render : function(context) {
        function in_context(value, context, is_safe) {
            if (value == undefined)
                return "";
            if (context.autoescape) {
                if (value instanceof Thistle.SafeString || value.is_safe || is_safe)
                    return value;
                return new Thistle.SafeString(value.toString().replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;').replace(/"/g, '&quot;').replace(/'/g, '&#39;'));
            }
            return value;
        };

        var output = this.filter_expression.resolve(context);

        return in_context(output, context);
    }
});

module.exports = {
    Node         : Node,
    NodeList     : NodeList,
    TextNode     : TextNode,
    VariableNode : VariableNode,
};

File: py/thistle/node.py
from .safestring import SafeData, EscapeData, SafeString, mark_safe, force_unicode, escape

__all__ = ['Node', 'NodeList', 'VariableNode', 'TextNode']

def localize(x, use_l10n=False): return x

class Node(object):
    # Set this to True for nodes that must be first in the template (although
    # they can be preceded by text nodes.
    must_be_first = False
    child_nodelists = ('nodelist',)

    def render(self, context):
        "Return the node rendered as a string"
        pass

    def __iter__(self):
        yield self

    def get_nodes_by_type(self, nodetype):
        "Return a list of all nodes (within this node and its nodelist) of the given type"
        nodes = []
        if isinstance(self, nodetype):
            nodes.append(self)
        for attr in self.child_nodelists:
            nodelist = getattr(self, attr, None)
            if nodelist:
                nodes.extend(nodelist.get_nodes_by_type(nodetype))
        return nodes

class NodeList(list):
    # Set to True the first time a non-TextNode is inserted by
    # extend_nodelist().
    contains_nontext = False

    def render(self, context):
        bits = []
        for node in self:
            if isinstance(node, Node):
                bits.append(self.render_node(node, context))
            else:
                bits.append(node)
        return mark_safe(''.join([force_unicode(b) for b in bits]))

    def get_nodes_by_type(self, nodetype):
        "Return a list of all nodes of the given type"
        nodes = []
        for node in self:
            nodes.extend(node.get_nodes_by_type(nodetype))
        return nodes

    def render_node(self, node, context):
        return node.render(context)

class TextNode(Node):
    def __init__(self, s):
        self.s = s

    def __repr__(self):
        return "<Text Node: '%s'>" % smart_str(self.s[:25], 'ascii',
                errors='replace')

    def render(self, context):
        return self.s

def _render_value_in_context(value, context):
    """
    Converts any value to a string to become part of a rendered template. This
    means escaping, if required, and conversion to a unicode object. If value
    is a string, it is expected to have already been translated.
    """
    value = localize(value, use_l10n=context.use_l10n)
    value = force_unicode(value)
    if (context.autoescape and not isinstance(value, SafeData)) or isinstance(value, EscapeData):
        return mark_safe(value.replace('&', '&amp;').replace('<', '&lt;').replace('>', '&gt;').replace('"', '&quot;').replace("'", '&#39;'))
    return value

class VariableNode(Node):
    def __init__(self, filter_expression):
        self.filter_expression = filter_expression

    def __repr__(self):
        return "<Variable Node: %s>" % self.filter_expression

    def render(self, context):
        try:
            output = self.filter_expression.resolve(context)
        except UnicodeDecodeError:
            # Unicode conversion can fail sometimes for reasons out of our
            # control (e.g. exception rendering). In that case, we fail quietly.
            return ''
        return _render_value_in_context(output, context)
